pkg/cache/cache_test.go

```diff
@@ -378,3 +382,63 @@ func TestCachePrecedence(t *testing.T) {
 	}
 
 }
+
+func TestRoleArnComposition(t *testing.T) {
+	role := "s3-reader"
+	audience := "sts.amazonaws.com"
+	tokenExpiration := "3600"
+	composeRoleArn := true
+	region := "test"
+	accountID := "111122223333"
+	resource := fmt.Sprintf("role/%s", role)
+
+	testSA := &v1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "default",
+			Namespace: "default",
+			Annotations: map[string]string{
+				"eks.amazonaws.com/role-arn":         role,
+				"eks.amazonaws.com/token-expiration": tokenExpiration,
+			},
+		},
+	}
+
+	testIdentity := ec2metadata.EC2InstanceIdentityDocument{
+		Region:    region,
+		AccountID: accountID,
+	}
+	metadataClient := NewFakeMetadataClient(false, testIdentity)
+
+	fakeClient := fake.NewSimpleClientset(testSA)
+	informerFactory := informers.NewSharedInformerFactory(fakeClient, 0)
+	informer := informerFactory.Core().V1().ServiceAccounts()
+
+	cache := New(audience, "eks.amazonaws.com", true, composeRoleArn, 86400, informer, nil, metadataClient)
+	cache.(*serviceAccountCache).hasSynced = func() bool { return true }
+	stop := make(chan struct{})
+	informerFactory.Start(stop)
+	cache.Start(stop)
+	defer close(stop)
+
+	err := wait.ExponentialBackoff(wait.Backoff{Duration: 10 * time.Millisecond, Factor: 1.0, Steps: 3}, func() (bool, error) {
+		return len(fakeClient.Actions()) != 0, nil
+	})
+	if err != nil {
+		t.Fatalf("informer never called client: %v", err)
+	}
+
+	err = wait.ExponentialBackoff(wait.Backoff{Duration: 10 * time.Millisecond, Factor: 1.0, Steps: 3}, func() (bool, error) {
+		return len(cache.(*serviceAccountCache).saCache) != 0, nil
+	})
+	if err != nil {
+		t.Fatalf("cache never called addSA: %v", err)
+	}
+
+	roleArn, _, _, _ := cache.Get("default", "default")
```

Comment:
This section can be written as:
```suggestion
	stop := make(chan struct{})
	informerFactory.Start(stop)
	cache.Start(stop)
	defer close(stop)

	var roleArn string
	err := wait.ExponentialBackoff(wait.Backoff{Duration: 10 * time.Millisecond, Factor: 1.0, Steps: 3}, func() (bool, error) {
		roleArn, _, _, _ = cache.Get("default", "default")
		return roleArn != "", nil
	})
	if err != nil {
		t.Fatalf("cache never returned role arn %v", err)
	}

```